### PR TITLE
Code coverage and codecov.io integration

### PR DESCRIPTION
**Pull request by tpluscode**
## Code coverage 

I switched to cake for build scripting in order to integrate code coverage (dotCover) and also simplified AppVeyor.yml a little bit in the process

Currently coverage is just above 48%. I'd hoped for more but that's what we got. 
The downside is that the CI build takes significantly longer, especially when executing the performance/large graph tests. I would try excluding some of those tests from coverage analysis as they most likely do no contribute (multiple same tests with datasets fo various sizes run through exact same code)

### PR status check

like-a-boss-decal

@kal, some time ago I requested access to GitHub for codecov.io. Alternatively maybe you could log in there and authorise the app so that we can get nice PR status check like below:

screen shot 2018-11-19 at 10 22 33

Once integrated, the status check will compare branch with master and also warn us if the coverage drops too much.

### CODE REVIEW

File: Libraries/dotNetRDF.Data.DataTables/dotNetRDF.Data.DataTables.csproj [hunk 1]
```diff
@@ -30,6 +30,8 @@
     <GenerateAssemblyInformationalVersionAttribute>false</GenerateAssemblyInformationalVersionAttribute>
     <CodeAnalysisRuleSet>..\..\dotnetrdf.ruleset</CodeAnalysisRuleSet>
     <RootNamespace>VDS.RDF.Data.DataTables</RootNamespace>
+    <DebugType>full</DebugType>
+    <DebugSymbols>true</DebugSymbols>
```
Comment (tpluscode) [inline Libraries/dotNetRDF.Data.DataTables/dotNetRDF.Data.DataTables.csproj]: I added full debug symbols for my first OpenCover attempt. Not sure if it's required for dotCover...

File: build.cake [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+Task("Codecov")
+    .Does(() => {
+        Codecov("coverage\\cobertura.xml");
+    });
+
+Task("TestNet452")
+    .IsDependentOn("Build")
+    .Does(RunTests(unitTests, "net452"))
+    .Does(RunTests(unitTests, "net452", "Category=fulltext"))
```
Comment (tpluscode) [inline build.cake]: Cannot remember why, but full text have to be executed separately from the rest (on AppVeyor at least).

File: build.cake [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+Task("Test")
+    .IsDependentOn("TestNet452")
+    .IsDependentOn("TestNetCore20")
+    .IsDependentOn("TestNetCore11")
+    .Does(() => { });
+
+Task("Cover")
+    .IsDependentOn("TestNetCore20")
+    .IsDependentOn("TestNetCore11")
```
Comment (tpluscode) [inline build.cake]: Coverage only runs for .NET 4.5.2 target to save some time as it's the platform with most complete feature set.

File: build.cake [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    .Does(() => {
+        if (DirectoryExists("coverage"))
+            CleanDirectories("coverage");
+    })
+    .Does(DotCover("net452", RunTests(unitTests, "net452")))
+    .Does(DotCover("fulltext", RunTests(unitTests, "net452", "Category=fulltext")))
```
Comment (tpluscode) [inline build.cake]: Ditto, fulltext tests executed separately

File: build.cake [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+          "./coverage/merged.dcvr",
+          "./coverage/dotcover.xml",
+          new DotCoverReportSettings {
+            ReportType = DotCoverReportType.DetailedXML,
+          });
+    })
+    .Does(() => {
+        StartProcess(
+          @".\tools\ReportGenerator.4.0.4\tools\net47\ReportGenerator.exe",
+          @"-reports:.\coverage\dotcover.xml -targetdir:.\coverage -reporttypes:Cobertura -assemblyfilters:-xunit*;-dotNetRDF.*Test");
```
Comment (tpluscode) [inline build.cake]: I submitted PR cake-build/cake#2360 to get Report Generator cake updated. Once it's merge we'll use the cake DSL

File: build.cake [hunk 6]
```diff
@@ -81,7 +81,8 @@ Task("Cover")
         if (DirectoryExists("coverage"))
             CleanDirectories("coverage");
     })
-    .Does(DotCover("net452", RunTests(unitTests, "net452")))
+    .Does(RunTests(unitTests, "net452", "Coverage=Skip"))
+    .Does(DotCover("net452", RunTests(unitTests, "net452", "Category!=explicit & Coverage!=Skip")))
```
Comment (tpluscode) [inline build.cake]: Excluding a handful more slowest tests (>10 seconds duration) shaves off 20 minutes from the AppVeyor run and only 0,12% coverage.

File: appveyor.yml [hunk 7]
```diff
@@ -15,40 +15,21 @@ install:
   - cmd: 7z x SHFBInstaller.zip -y
   - ps: msiexec /i InstallResources\SandcastleHelpFileBuilder.msi /quiet /qn /norestart /log install.log
 
-
 before_build:
-  - dotnet restore dotNetRDF.sln
```
Comment (kal) [inline appveyor.yml]: Is this now handled by Cake?

Comment (tpluscode) [reply appveyor.yml]: Indirectly. `dotnet build` actually restores packages automatically

File: build.cake [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+            NoBuild = true,
+            MSBuildSettings = new DotNetCoreMSBuildSettings()
+        };
+
+        settings.MSBuildSettings.Properties["version"] = new [] { version };
+
+        DotNetCorePack(path.FullPath, settings);
+    });
+
+Task("GitVersion")
```
[the remark this answers is not in the document]
Comment (kal) [reply build.cake]: Ah yes, I hadn't twigged that the GitVersion in cake only runs on a local build. I think that's Good Enough :-)

**Comment by kal** [comment]
Hi @tpluscode - I think that the default configuration on codecov is monitoring the develop branch. I guess we should switch this to master?

**Comment by tpluscode** [comment]
Don't know why it says so. We don't even have a develop branch. According to the docs, it should use repo's default branch. I think it will sort itself out once there is a coverage report submitted for master.

**Comment by kal** [comment]
:+1: Lets merge it and give it a go then :)

**Comment by tpluscode** [comment]
Hm, still master. What is the default branch setting on GH?

**Pull request merged**